src/core/remote_io.c

```diff
@@ -0,0 +1,688 @@
+    }
+
+    struct remote_io_request_event* event = remote_io_pop_request_event(device);
+    if (event == NULL) {
+        return ret;
+    }
+
+    if (remote_io_get_request_state(event->cpu_id, event->vcpu_id) != REMOTE_IO_STATE_PENDING) {
+        objpool_free(&remote_io_request_event_pool, event);
+        return ret;
+    }
+
+    /** Calculate the remaining number of pending I/O requests */
+    ret = (long int)remote_io_get_request_event_count(device);
+
+    remote_io_set_request_state(event->cpu_id, event->vcpu_id, REMOTE_IO_STATE_PROCESSING);
+
+    struct remote_io_request* request = remote_io_get_request(event->cpu_id, event->vcpu_id);
+
+    /** Write the I/O request information to the backend VM's vCPU registers */
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(0), request->addr);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(1), request->op);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(2), request->value);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(3), request->access_width);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(4), event->cpu_id);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(5), event->vcpu_id);
+
+    objpool_free(&remote_io_request_event_pool, event);
+
+    return ret;
+}
+
+/**
+ * @brief Handles the Remote I/O read and write operations
+ * @param value Value to be written or read
+ * @param cpu_id CPU ID of the frontend VM that issued the I/O request
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns true if the operation was successful, false otherwise
+ */
+static bool remote_io_handle_rw(unsigned long value, unsigned long cpu_id, unsigned long vcpu_id)
+{
+    if (remote_io_get_request_state(cpu_id, vcpu_id) != REMOTE_IO_STATE_PROCESSING) {
+        return false;
+    }
+    remote_io_set_request_value(cpu_id, vcpu_id, value);
+    remote_io_set_request_state(cpu_id, vcpu_id, REMOTE_IO_STATE_COMPLETE);
+    return true;
+}
+
+/**
+ * @brief Performs the post work after the completion of the I/O request
+ * @note This function is executed by the frontend VM and is responsible for updating the
+ *       vCPU register in case of a read operation and activating the frontend vCPU
+ * @param event Message event (REMOTE_IO_CPU_MSG_*)
+ * @param cpu_id CPU ID of the frontend VM that issued the I/O request
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns true if the operation was successful, false otherwise
+ */
+static bool remote_io_cpu_post_work(uint32_t event, uint8_t cpu_id, uint8_t vcpu_id)
+{
+    if (remote_io_get_request_state(cpu_id, vcpu_id) != REMOTE_IO_STATE_COMPLETE) {
+        return false;
+    }
+    struct remote_io_request* request = remote_io_get_request(cpu_id, vcpu_id);
+
+    switch (event) {
+        case REMOTE_IO_CPU_MSG_READ:
+            vcpu_writereg(cpu()->vcpu, request->reg, request->value);
+            break;
+        default:
+            break;
+    }
```

Comment:
Since the I/O request slot is managed by a finite state machine (FSM) and updated sequentially by both the frontend and backend VMs, locking is unnecessary. To break it down:

1. The request is created by the frontend VM during the MMIO emulation handler (`remote_io_mmio_emul_handler`).

From that point on, the request slot is updated sequentially by the frontend and backend VMs. Specifically:

2. During the ask operation (`remote_io_handle_ask`), the backend updates the request state.
3. Once the backend finishes processing the request, it updates both the state and value (`remote_io_handle_rw`).
4. Finally, the frontend VM updates the state during the post-work phase (`remote_io_cpu_post_work`).

This ensures that:

1. The backend and frontend VMs will never access or modify the request simultaneously. When the backend VM operates in polling mode, it checks for pending requests by popping a request event, which is protected by the list lock, rather than directly inspecting the request. In interrupt mode, the interrupt is only injected after the I/O request slot has been fully updated. As in polling mode, the backend checks for new pending requests by examining the event list. Only when an event is present (indicating that the frontend has already inserted the request into the I/O request array via `remote_io_push_request_event` after `remote_io_insert_request`) will the backend fetch the I/O request slot.
2. If the frontend VM runs on multiple vCPUs, each vCPU can only make one request at a time. Since the vCPU goes into an idle state after issuing a request, it is impossible for multiple vCPUs to access the same slot simultaneously.
3. If the backend VM runs on multiple vCPUs, each vCPU handling the hypercall will be responsible for processing only one request slot at a time, given by the CPU and vCPU IDs.

Although I have tested the infrastructure on real hardware platforms (e.g., ZCU102), running both frontend and backend VMs on multiple vCPUs without encountering any issues, please feel free to highlight any potential corner cases that I may have missed at this stage.

Reply:
Thanks for the explanation I think I understand
However, I fail to understand in what scenarios is the lock actually needed if the vcpu will always for a request to be completed before issuing another

Reply:
Yes, you are correct! A global lock is unnecessary, as in the ask routine, we only fetch the I/O slot from the array when there are pending I/O requests in the event request list. In fact, I realized this while drafting the message above. :)

In earlier versions, I implemented the lock because there was no event request list; therefore, the verification and the actual retrieval process were performed on the same object.